Source/WebCore/loader/ContentFilter.cpp

```diff
@@ -123,6 +123,73 @@ bool ContentFilter::continueAfterWillSendRequest(ResourceRequest& request, const
     return !request.isNull();
 }
 
+ContentFilter::ContentFilterCallbackAggregator::~ContentFilterCallbackAggregator()
+{
+    ASSERT(m_callback);
+    CheckedPtr contentFilter = m_contentFilter;
+
+    if (m_blocked) {
+        contentFilter->didDecide(State::Blocked);
+        m_callback(ResourceRequest());
+        return;
+    }
+    if (m_platformFiltersAllowed == contentFilter->m_contentFilters.size())
+        contentFilter->didDecide(State::Allowed);
+    m_callback(WTFMove(m_request));
+}
+
+void ContentFilter::ContentFilterCallbackAggregator::didReceivePlatformContentFilterDecision(PlatformContentFilter& platformContentFilter, ResourceRequest&& request, PlatformContentFilter::State state)
+{
+    CheckedPtr contentFilter = m_contentFilter;
+
+    if (state == PlatformContentFilter::State::Allowed)
+        ++m_platformFiltersAllowed;
+
+    bool blocked = state == PlatformContentFilter::State::Blocked;
+    if (blocked && !m_blocked) {
+        m_blocked = true;
+        contentFilter->m_blockingContentFilter = platformContentFilter;
+        return;
+    }
+
+    m_request = WTFMove(request);
+}
+
+ContentFilter::ContentFilterCallbackAggregator::ContentFilterCallbackAggregator(ContentFilter& contentFilter, CompletionHandler<void(ResourceRequest&&)>&& callback)
+    : m_contentFilter(contentFilter)
+    , m_callback(WTFMove(callback))
+{
+}
+
+void ContentFilter::continueAfterWillSendRequest(ResourceRequest&& request, const ResourceResponse& redirectResponse, CompletionHandler<void(ResourceRequest&&)>&& completionHandler)
+{
+    ASSERT(RunLoop::isMain());
+
+    // Return early if no platform content filters are enabled.
+    if (!anyContentFilterIsEnabled()) {
+        continueAfterWillSendRequest(request, redirectResponse);
+        completionHandler(WTFMove(request));
+        return;
+    }
+
+    LOG(ContentFiltering, "ContentFilter received request for <%{sensitive}s> with redirect response from <%{sensitive}s>.\n", request.url().string().utf8().data(), redirectResponse.url().string().utf8().data());
+    ASSERT(m_state == State::Stopped || m_state == State::Filtering);
+
+    Ref contentFilterCallbackAggregator = ContentFilterCallbackAggregator::create(*this, WTFMove(completionHandler));
+
+    for (Ref platformContentFilter : m_contentFilters) {
+        CompletionHandler<void(ResourceRequest, PlatformContentFilter::State)> completion = [contentFilterCallbackAggregator, platformContentFilter] (ResourceRequest&& request, PlatformContentFilter::State state) mutable {
+            ASSERT(RunLoop::isMain());
+            contentFilterCallbackAggregator->didReceivePlatformContentFilterDecision(platformContentFilter, WTFMove(request), state);
+        };
+
+        if (platformContentFilter->needsMoreData()) {
+            auto requestCopy = request;
+            platformContentFilter->willSendRequest(WTFMove(requestCopy), redirectResponse, WTFMove(completion));
+        }
+    };
```

Comment:
Why the semicolon here? Maybe I'm misreading the diff but it look like the end of the `for` loop.